.tekton/retag-image-task.yaml

```diff
@@ -0,0 +1,108 @@
+  - name: INPUT_IMAGE_REPO
+    description: Repository of the input image.
+    type: string
+  - name: INPUT_IMAGE_TAG
+    description: Tag of the input image.
+    type: string
+  - name: OUTPUT_IMAGE_REPO
+    description: Repository of the output image.
+    type: string
+  - name: OUTPUT_IMAGE_TAG
+    description: Repository of the output image.
+    type: string
+  results:
+  - name: IMAGE_DIGEST
+    description: Digest of the output image (will be the same as of the input one).
+  - name: IMAGE_URL
+    description: Image repository and tag of the output image.
+  - name: IMAGE_REF
+    description: Image reference of the output image containing both the repository, the tag and the digest.
+  steps:
+  - name: retag-image
+    image: quay.io/konflux-ci/release-service-utils:latest@sha256:0fdaefe6748962003e24f3faede4edf1abbea63112c6ad973654a0726aafb9b8
+    script: |
+      #!/usr/bin/env bash
+      set -euo pipefail
+
+      SKOPEO_RETRIES=5
+
+      function copy_image() {
+        local -r input_url="${1}"
+        local -r output_url="${2}"
+
+        if [[ "${input_url}" == "${output_url}" ]]; then
+          >&2 echo "Will not copy image ${input_url} into itself."
+          exit 5
+        fi
+
+        echo "Copying image from ${input_url} ..."
+        echo "... to ${output_url} ..."
+
+        # cosign copies not just cosign artifacts but also images. It understands and copies index images.
+        # --force argument is needed to prevent the command failing when it runs into (partial) previous copies with
+        # errors like the following.
+        #   Error: image "quay.io/rhacs-eng/collector:sha256-4b509fdf27761150a5a5ce519f087bc6d69af9d886a63f67f70b1e293643ea19" already exists. Use `-f` to overwrite
+        cosign copy --force "${input_url}" "${output_url}"
+
+        oras cp --recursive "${input_url}" "${output_url}"
+      }
+
+      function get_digest() {
+        local -r image_url="${1}"
+        skopeo inspect \
+          --retry-times="${SKOPEO_RETRIES}" \
+          --format '{{.Digest}}' \
+          --no-tags \
+          "docker://${image_url}"
+      }
+
+      INPUT_URL="$(params.INPUT_IMAGE_REPO):$(params.INPUT_IMAGE_TAG)"
+      OUTPUT_URL="$(params.OUTPUT_IMAGE_REPO):$(params.OUTPUT_IMAGE_TAG)"
+
+      # This copies the images themselves.
+      # It also handles the case when the image is index on top of arch-specific containers in which case arch-specific
+      # containers are also copied (but not tags of arch-specific containers if they were additionally tagged).
+      copy_image "${INPUT_URL}" "${OUTPUT_URL}"
+
+      echo "Figuring the input image digest ..."
+      INPUT_DIGEST="$(get_digest "${INPUT_URL}")"
+
+      echo "Figuring the output image digest ..."
+      OUTPUT_DIGEST="$(get_digest "${OUTPUT_URL}")"
+
+      echo "Checking if the input image digest ${INPUT_DIGEST} is the same as the output image digest ${OUTPUT_DIGEST} ..."
+      [[ "${INPUT_DIGEST}" == "${OUTPUT_DIGEST}" ]]
```

Comment:
> An error exit

Since the script begins with `set -e`, it will exit with code `1` should the expression in `[[` `]]` evaluate to false.

> Under which condition would that happen?

It should not happen but I don't trust `cosign` enough since it's a new tool to me (or `skopeo copy` which I replaced with `cosign copy`).
Having digests match is fundamental proof that we did not modify the image and it's bit-by-bit identical that's why I wanted to add this check in the task instead of validating that one-off during the work on this PR.

Also, this check documents at the code level that digests can be used interchangeably.